### PR TITLE
Introducing PVC list and PVC mapping for Transfers

### PR DESCRIPTION
**Pull request by pranavgaikwad**
This PR introduces PVC list to be passed as input when creating Transfers. All transfers take the list as input and consume it based on their current capabilities:
1. Rsync Transfer cannot process PVCs belonging to multiple source/destination namespaces
2. Rclone Transfer cannot process multiple PVCs

### CODE REVIEW

File: state_transfer/transfer/pvc.go [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+)
+
+type PersistentVolumeClaimList []PVC
+
+type PVC interface {
+	Source() PersistentVolumeClaim
+	Destination() PersistentVolumeClaim
+}
+
+type PersistentVolumeClaim interface {
```
Comment (alaypatel07) [inline state_transfer/transfer/pvc.go:19]: users will wonder what is the difference between this type and `corev1.PersistentVolumeClaim`

Can we rename this as `LabelSafePersistentVolumeClaim` or `ValidatedPersistentVolumeClaim`? Although the latter could be ambiguous in what the `Validation` is.

File: state_transfer/transfer/pvc.go [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+	"encoding/hex"
+	"fmt"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+type PersistentVolumeClaimList []PVC
+
+type PVC interface {
```
Comment (alaypatel07) [inline state_transfer/transfer/pvc.go:14]: rename to `PVCMapping` or `PVCPair` ?

File: state_transfer/transfer/pvc.go [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+import (
+	"crypto/md5"
+	"encoding/hex"
+	"fmt"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+type PersistentVolumeClaimList []PVC
```
Comment (alaypatel07) [inline state_transfer/transfer/pvc.go:12]: rename to `PVCMappingList`?

File: state_transfer/transfer/pvc.go [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+	Source() PersistentVolumeClaim
+	Destination() PersistentVolumeClaim
+}
+
+type PersistentVolumeClaim interface {
+	Claim() *v1.PersistentVolumeClaim
+	ValidatedName() string
+}
+
+type pvc struct {
```
Comment (alaypatel07) [inline state_transfer/transfer/pvc.go:24]: NIT: can we move the implementation into a separate file?

File: state_transfer/transfer/rsync/validation.go [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+	validatedPVCNameMaxLength = 63
+	pvcNameBadCharacters      = `[\\.]+`
+)
+
+// validatePVCList validates list of PVCs provided to rsync transfer
+func validatePVCList(pvcList transfer.PersistentVolumeClaimList) error {
+	srcNamespaces := pvcList.GetSourceNamespaces()
+	destNamespaces := pvcList.GetDestinationNamespaces()
+	if len(srcNamespaces) > 1 || len(destNamespaces) > 1 {
+		return fmt.Errorf("rsync transfer does not support migrating PVCs belonging to multiple source/destination namespaces")
```
Comment (alaypatel07) [inline state_transfer/transfer/rsync/validation.go:21]: +1, I like this validation is in rsync package.

What we are saying here is that the interfaces themselves dont put the limitation of a single namespace, but rsync implementation does, and we can change that in the future as needed.

File: state_transfer/transfer/rsync/validation.go [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+				errorsutil.NewAggregate([]error{
+					fmt.Errorf("pvc name validation failed for pvc %s with error", pvc.Source().Claim().Name),
+					err,
+				}))
+		}
+	}
+	return errorsutil.NewAggregate(validationErrors)
+}
+
+// validatePVCName validates pvc names for rsync transfer
```
Comment (alaypatel07) [inline state_transfer/transfer/rsync/validation.go:40]: Can you add the following to the comments?

```
Users can provide their own implementation of PVC interfaces, this avoids breaking rsync implementation if the user-provided implementation is broken.
```

**Review by alaypatel07 — CHANGES_REQUESTED** [review]
This looks awesome. It:
1. provides the interfaces in higher-level packages
2. provides a default implementation
3. checks for validation in rsync package, avoids running into errors when users provide non-default implementation

Added some suggestions on naming, with the main rationale being avoiding confusion with core types

**Review by shawn-hurley — APPROVED** [review]
once we address the naming concerns from Alay other than that I agree +1 LGTM

**Review by alaypatel07 — APPROVED** [review]
LGTM

**Pull request merged**